# Pokémon Statistics

The Pokémon media franchise is well known as an international sensation. It all began with two humble video games for the original Game Boy developed by Game Freak and published by Nintendo in 1996. From this first generation with its 151 Pokémon, the Pokémon video games have continued to be developed and expanded for over twenty years. As of writing, there are now seven generations with a total of 802 Pokémon.

Alberto Barradas published a [dataset](https://www.kaggle.com/abcsds/pokemon) of the video game statistics of the 721 Pokémon of the first six generations, providing a convenient toy data project for exercises in data manipulation and visualization. This notebook explores the dataset for opportunities for such practice.

<a id='0'></a>

### Table of Contents
- [Part 1: Exploratory Analysis](#Part 1)
    - [1. Dataset](#1)
    - [2. Base Stat Rankings](#2)
        - [2.1 Total](#2.1)
        - [2.2 HP (Damage Sponge)](#2.2)
        - [2.3 Attack and Sp. Atk (Sweepers)](#2.3)
        - [2.4 Defense and Sp. Def (Walls](#2.4)
        - [2.5 Speed (Speedsters)](#2.5)
- [Part 2: Classification](#Part 2)
    - [3. Competitive Smogon Tiers](#3)

## Part 1: Exploratory Analysis

### Library Imports

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

[Return to Top](#0)

<a id='1'></a>
### 1. Dataset

In [2]:
pk = pd.read_csv('Pokemon.csv')

In [3]:
pk.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation Legendary  
0       65       65     45           1     False  
1       80       80     60           1     False  
2      100      100     80           1     False  
3      122      120     80           1     False  
4       60       50     65           1     False

In [6]:
pk.tail()

#                 Name   Type 1 Type 2  Total  HP  Attack  Defense  \
795  719              Diancie     Rock  Fairy    600  50     100      150   
796  719  DiancieMega Diancie     Rock  Fairy    700  50     160      110   
797  720  HoopaHoopa Confined  Psychic  Ghost    600  80     110       60   
798  720   HoopaHoopa Unbound  Psychic   Dark    680  80     160       60   
799  721            Volcanion     Fire  Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation Legendary  
795      100      150     50           6      True  
796      160      110    110           6      True  
797      150      130     70           6      True  
798      170      130     80           6      True  
799      130       90     70           6      True

In the Pokémon video games, Pokémon battle one another. The battle mechanics are such that calculations are made based on base statistics of the Pokémon species. There are numerous other modifiers, such as level, attack statistics, and other factors, but these statistics are outside the scope of this dataset.

The base statistics as given in this data set are:
- #: ID number (National Pokédex number in-game) of the pokémon
- Name: The English name of the Pokémon (names vary between languages)
- Type 1: The type determines "elemental" weakness/resistance modifiers to attacks
- Type 2: Some Pokémon have a two types (called "dual-types")
- Total: The sum of the base stats HP, Attack, Defense, Sp. Atk, Sp. Def, and Speed.
- HP: Health Points, defining how much damage the Pokémon can withstand before fainting
- Attack: Base modifier for physical attacks
- Defense: Base modifier for resisting physical attacks
- Sp. Atk: Base modifier for special attacks
- Sp. Def: Base modifier for resisting special attacks
- Speed: Base modifier for determining attack order
- Generation: The Video Game Generation the Pokémon first appeared in (1-6)
- Legendary: Whether the pokémon is identified in-game a special "legendary" pokémon; Legendaries tend to very powerful, having above-average base statistics

In [4]:
pk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
#             800 non-null int64
Name          800 non-null object
Type 1        800 non-null object
Type 2        414 non-null object
Total         800 non-null int64
HP            800 non-null int64
Attack        800 non-null int64
Defense       800 non-null int64
Sp. Atk       800 non-null int64
Sp. Def       800 non-null int64
Speed         800 non-null int64
Generation    800 non-null int64
Legendary     800 non-null bool
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


We may note that although there are 720 unique pokémon, there are in fact 799 entries. This is due to the sixth generation's introduction of "Mega" modifiers, which change the base statistics but not the pokémon's ID number.

Clearly, there are 799 - 720 = 79 "Mega" variants and 800 - 414 = 386 entries with only one type.

[Return to Top](#0)

---

<a id='2'></a>

## 2. Base Stat Rankings

<a id='2.1'></a>

### 2.1 Totals

In [20]:
pk.sort_values(by='Total', ascending=False).head(n=10)

#                     Name   Type 1    Type 2  Total   HP  Attack  \
426  384    RayquazaMega Rayquaza   Dragon    Flying    780  105     180   
164  150      MewtwoMega Mewtwo Y  Psychic       NaN    780  106     150   
163  150      MewtwoMega Mewtwo X  Psychic  Fighting    780  106     190   
422  382      KyogrePrimal Kyogre    Water       NaN    770  100     150   
424  383    GroudonPrimal Groudon   Ground      Fire    770  100     180   
552  493                   Arceus   Normal       NaN    720  120     120   
712  646       KyuremWhite Kyurem   Dragon       Ice    700  125     120   
711  646       KyuremBlack Kyurem   Dragon       Ice    700  125     170   
409  373  SalamenceMega Salamence   Dragon    Flying    700   95     145   
413  376  MetagrossMega Metagross    Steel   Psychic    700   80     145   

     Defense  Sp. Atk  Sp. Def  Speed  Generation Legendary  
426      100      180      100    115           3      True  
164       70      194      120    140           1      True  
163      100      154      100    130           1      True  
422       90      180      160     90           3      True  
424      160      150       90     90           3      True  
552      120      120      120    120           4      True  
712       90      170      100     95           5      True  
711      100      120       90     95           5      True  
409      130      120       90    120           3     False  
413      150      105      110    110           3     False

By base stat totals, the three most powerful pokémon are the Mega variant of Rayquaza and the two Mega variants of Mewtwo. After them come the Mega variants of Kyogre and Groudon, while the "God" of the pokémon universe, Arceus, only comes in sixth in base stat totals.

The funny implications of this are that the Sky (represented by Rayquaza), Ground (represented by Groudon), and Ocean (represented by Kyogre), are more powerful than "God," and that in the pokémon universe one should perhaps worship the natural earth instead. Also potentiall funnier is that Mewtwo is a man-made pokémon in the pokémon universe, meaning that humanity essentially created a being more powerful than "God."

However, it is worth noting that all six of these are Legendary. If one wishes to apply this sort of information to for some applications for the video games, perhaps in some sort of competitive battling, it may be worthwhile to exclude Legendaries, which are often banned or restricted in regular play.

In [19]:
pk[pk['Legendary'] == False].sort_values(by='Total', ascending=False).head(n=10)

#                       Name  Type 1   Type 2  Total   HP  Attack  \
413  376    MetagrossMega Metagross   Steel  Psychic    700   80     145   
494  445      GarchompMega Garchomp  Dragon   Ground    700  108     170   
409  373    SalamenceMega Salamence  Dragon   Flying    700   95     145   
268  248    TyranitarMega Tyranitar    Rock     Dark    700  100     164   
313  289                    Slaking  Normal      NaN    670  150     160   
141  130      GyaradosMega Gyarados   Water     Dark    640   95     155   
283  260      SwampertMega Swampert   Water   Ground    635  100     150   
8      6  CharizardMega Charizard Y    Fire   Flying    634   78     104   
7      6  CharizardMega Charizard X    Fire   Dragon    634   78     130   
12     9    BlastoiseMega Blastoise   Water      NaN    630   79     103   

     Defense  Sp. Atk  Sp. Def  Speed  Generation Legendary  
413      150      105      110    110           3     False  
494      115      120       95     92           4     False  
409      130      120       90    120           3     False  
268      150       95      120     71           2     False  
313      100       95       65    100           3     False  
141      109       70      130     81           1     False  
283      110       95      110     70           3     False  
8         78      159      115    100           1     False  
7        111      130       85    100           1     False  
12       120      135      115     78           1     False

Once again, Mega variants dominate.

[Return to Top](#0)

---

<a id='2.2'></a>

### 2.2 HP (Damage Sponge)

In [10]:
pk.sort_values(by='HP', ascending=False).head(n=5)

#       Name   Type 1 Type 2  Total   HP  Attack  Defense  Sp. Atk  \
261  242    Blissey   Normal    NaN    540  255      10       10       75   
121  113    Chansey   Normal    NaN    450  250       5        5       35   
217  202  Wobbuffet  Psychic    NaN    405  190      33       58       33   
351  321    Wailord    Water    NaN    500  170      90       45       90   
655  594  Alomomola    Water    NaN    470  165      75       80       40   

     Sp. Def  Speed  Generation Legendary  
261      135     55           2     False  
121      105     50           1     False  
217       58     33           2     False  
351       45     60           3     False  
655       45     65           5     False

[Return to Top](#0)

---

<a id='2.3'></a>

### 2.3 Attack and Sp. Atk (Sweepers)

In [11]:
pk.sort_values(by=['Attack','Sp. Atk'], ascending=False).head(n=5)

#                     Name   Type 1    Type 2  Total   HP  Attack  \
163  150      MewtwoMega Mewtwo X  Psychic  Fighting    780  106     190   
232  214  HeracrossMega Heracross      Bug  Fighting    600   80     185   
426  384    RayquazaMega Rayquaza   Dragon    Flying    780  105     180   
429  386       DeoxysAttack Forme  Psychic       NaN    600   50     180   
424  383    GroudonPrimal Groudon   Ground      Fire    770  100     180   

     Defense  Sp. Atk  Sp. Def  Speed  Generation Legendary  
163      100      154      100    130           1      True  
232      115       40      105     75           2     False  
426      100      180      100    115           3      True  
429       20      180       20    150           3      True  
424      160      150       90     90           3      True

In [12]:
pk.sort_values(by=['Sp. Atk', 'Attack'], ascending=False).head(n=5)

#                   Name   Type 1  Type 2  Total   HP  Attack  Defense  \
164  150    MewtwoMega Mewtwo Y  Psychic     NaN    780  106     150       70   
426  384  RayquazaMega Rayquaza   Dragon  Flying    780  105     180      100   
429  386     DeoxysAttack Forme  Psychic     NaN    600   50     180       20   
422  382    KyogrePrimal Kyogre    Water     NaN    770  100     150       90   
71    65  AlakazamMega Alakazam  Psychic     NaN    590   55      50       65   

     Sp. Atk  Sp. Def  Speed  Generation Legendary  
164      194      120    140           1      True  
426      180      100    115           3      True  
429      180       20    150           3      True  
422      180      160     90           3      True  
71       175       95    150           1     False

[Return to Top](#0)

---

<a id='2.4'></a>

### 2.4 Defense and Sp. Def (Walls)

In [22]:
pk.sort_values(by=['Defense', 'Sp. Def'], ascending=False).head(n=5)

#                 Name Type 1  Type 2  Total  HP  Attack  Defense  \
230  213              Shuckle    Bug    Rock    505  20      10      230   
224  208  SteelixMega Steelix  Steel  Ground    610  75     125      230   
333  306    AggronMega Aggron  Steel     NaN    630  70     140      230   
414  377             Regirock   Rock     NaN    580  80     100      200   
223  208              Steelix  Steel  Ground    510  75      85      200   

     Sp. Atk  Sp. Def  Speed  Generation Legendary  
230       10      230      5           2     False  
224       55       95     30           2     False  
333       60       80     50           3     False  
414       50      100     50           3      True  
223       55       65     30           2     False

In [23]:
pk.sort_values(by=['Sp. Def', 'Defense'], ascending=False).head(n=5)

#                 Name   Type 1  Type 2  Total   HP  Attack  Defense  \
230  213              Shuckle      Bug    Rock    505   20      10      230   
415  378               Regice      Ice     NaN    580   80      50      100   
430  386  DeoxysDefense Forme  Psychic     NaN    600   50      70      160   
422  382  KyogrePrimal Kyogre    Water     NaN    770  100     150       90   
269  249                Lugia  Psychic  Flying    680  106      90      130   

     Sp. Atk  Sp. Def  Speed  Generation Legendary  
230       10      230      5           2     False  
415      100      200     50           3      True  
430       70      160     90           3      True  
422      180      160     90           3      True  
269       90      154    110           2      True

[Return to Top](#0)

---

<a id='2.5'></a>

### 2.5 Speed (Speedsters)

In [21]:
pk.sort_values(by=['Speed'], ascending=False).head(n=5)

#                       Name   Type 1  Type 2  Total  HP  Attack  \
431  386          DeoxysSpeed Forme  Psychic     NaN    600  50      95   
315  291                    Ninjask      Bug  Flying    456  61      90   
428  386         DeoxysNormal Forme  Psychic     NaN    600  50     150   
154  142  AerodactylMega Aerodactyl     Rock  Flying    615  80     135   
71    65      AlakazamMega Alakazam  Psychic     NaN    590  55      50   

     Defense  Sp. Atk  Sp. Def  Speed  Generation Legendary  
431       90       95       90    180           3      True  
315       45       50       50    160           3     False  
428       50      150       50    150           3      True  
154       85       70       95    150           1     False  
71        65      175       95    150           1     False

[Return to Top](#0)

---

<a id='Part 2'></a>
## Part 2: Classification

### 3. Competitive Smogon Tiers

The Smogon competitive Pokémon battling community